sklearn/naive_bayes.py

```diff
@@ -117,6 +119,320 @@ def predict_proba(self, X):
+    >>> X = np.array([[1.5, 2.3, 5.7, 0, 1],
+    ...               [2.7, 3.8, 2.3, 1, 0],
+    ...               [1.7, 0.1, 4.5, 1, 0]])
+    >>> y = np.array([1, 0, 0])
+    >>> clf = GeneralNB(
+    ...         [("gaussian", GaussianNB(), [0, 1, 2]),
+    ...          ("categorical", CategoricalNB(), [3, 4])])
+    >>> clf.fit(X, y)
+    GeneralNB(models=[('gaussian', GaussianNB(...), [0, 1, 2]),
+                      ('categorical', CategoricalNB(...), [3, 4])])
+    >>> clf.predict(X[:1,])
+    array([1])
+    """
+
+    def __init__(self, models):
+        self.models = models
+        self.models_ = None
+        self.classes_ = None
+        self.n_features_ = None
+        self._cols = None
+        self._df_cols = None
+        self._is_fitted = False
+
+    def fit(self, X, y):
+        """Fit X and y to the naive Bayes estimators.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training sample, where `n_samples` is the number of samples
+            and `n_features` is the number of features.
+        y : array-like, shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+        self._validate_models(X)
+        self._check_X_y(X, y)
+
+        self.classes_ = np.unique(y)
+
+        self.models_ = [
+            (name, nb_model.fit(_safe_indexing(X, cols, axis=1), y), cols)
+            for (name, nb_model, _), cols
+            in zip(self.models, self._cols)]
+
+        self._is_fitted = True
+
+        return self
+
+    def _joint_log_likelihood(self, X):
+        """Calculate the posterior log probability of sample X
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training sample, where `n_samples` is the number of samples
+            and `n_features` is the number of features.
+
+        Returns
+        -------
+        jll : ndarray, shape (1, n_classes)
+            Posterior log probability.
+
+        Raises
+        ------
+        NotFittedError
+            If estimators have not been fitted
+        """
+        if not self._is_fitted:
+            raise NotFittedError("Call the fit() method first "
+                                 "before calling predict().")
+
+        # Obtain the log priors from each fitted estimator
+        all_log_priors = [
+            nb_model.class_log_prior_
+            if hasattr(nb_model, 'class_log_prior_')
+            else np.log(nb_model.class_prior_)
+            for _, nb_model, _ in self.models_]
+
+        # Ensure class log priors are the same for all estimators
+        all_log_priors = np.hstack([all_log_priors])
+        if np.max(np.ptp(all_log_priors, axis=0)) < 1e-6:
+            log_prior = all_log_priors[0]
+        else:
+            raise ValueError("Class priors for every estimator "
+                             "must be the same.")
```

Comment:
Please move this check to `fit`, if we need it at all?

Reply:
Okay will move to `fit`. I feel it's necessary just in case the user specifies different class priors for the models.